### PR TITLE
Add benchmark for ppxlib driver

### PR DESCRIPTION
**Pull request by gridbugs**
This adds a target `make bench` which measures the time taken to run the ppxlib driver on some input files. I've added two drivers to be benchmarked:
 - one which applies ppx_sexp_conv (which is added as a vendored package)
 - one which does nothing

Each driver specifies some input files copied from MIT licensed projects. The benchmark runner invokes each driver on each of its specified inputs and measures the mean and stddev of the time it took to run each driver on each of its inputs.

### CODE REVIEW

**Comment by tmattio** [comment]
This is awesome, thanks @gridbugs!! 🎉 

I'm curious to know if we could integrate it with [current-bench](https://github.com/ocurrent/current-bench)? (that would allow us to have benchmark reports in GitHub PRs automatically IIUC?)

Do you know how easy that would be @shakthimaan?

**Comment by gridbugs** [comment]
Thanks @tmattio! I forgot to mention in the description, but integrating with current-bench is the goal for this.

**Comment by gridbugs** [comment]
This fails to build on 4.04 and 4.05 because it uses standard library functions that didn't exist in those versions of the language. Also, the pinned version of ppx_sexp_conv is too new to be built by those versions of the compiler. Is it making sure that the benchmarks build on old versions of the compiler?

**Comment by panglesd** [comment]
There is actually another problem with not separating the benchmark in another package: the dependencies of the benchmark and the core library are mixed. This means that:
- `yojson` is added as a dependency while it is not useful to `ppxlib` itself
- Any time we want to add a new benchmark, we might add extra dependencies.

Since `ppxlib` is meant to be the "goto" library for all ppx rewriters, I think we should avoid as much as possible adding new dependencies. So I think that we should really go for solution 3.

If you do not have that much time to give to ppxlib, I could try to implement this solution myself. In any case, thanks a lot for the work! As @pitag-ha said, it is really great result!

**Comment by pitag-ha** [comment]
> the dependencies of the benchmark and the core library are mixed

To avoid that, @gridbugs has added it as a `with-test` dependency, not as a hard dependency. I myself don't have a preference between the two things: `with-test` dependency or separate opam package. I've also asked @kit-ty-kate if one of them would be better from an `opam` point of view and she's said that both are equally fine. So I'd say it's up to you, @gridbugs and @panglesd.

**Comment by panglesd** [comment]
> To avoid that, @gridbugs has added it as a with-test dependency

Yes, that's right, sorry I forgot about that!

So in fact, the [reasoning](https://github.com/ocaml-community/yojson/pull/117#issuecomment-983447660) for separating benchmark from the tests in yojson was that it can be time costly (and also, depending on your definition of test, it is not exactly a test as there are no exact failing/succeeding conclusion).

In this case, if the tests run sufficiently fast and are not computation-intensive, I'm fine with any solution!

**Comment by shakthimaan** [comment]
> I'm curious to know if we could integrate it with [current-bench](https://github.com/ocurrent/current-bench)?

Yes. Please follow the steps in `Enrolling a repository` section provided in the following blog post to add your project to current-bench:
https://tarides.com/blog/2021-08-26-benchmarking-ocaml-projects-with-current-bench

**Comment by pitag-ha** [comment]
> Yes. Please follow the steps in Enrolling a repository section provided in the following blog post

Yes, I'll do that as soon as this PR is merged.

Btw, im currently at the MirageOs retreat and won't have much time to review this and next week. If you have time, @panglesd , that'd be very cool. If not, I'll review as soon as I'm back.

**Comment by gridbugs** [comment]
I added a `ppxlib-bench` package to this repo with the extra deps needed for benchmarks. This package relies on a newer version of ocaml (4.10 rather than 4.04 due to its dependency on ppx_sexp_conv). This has resulted in ci only being run on versions of the compiler >= 4.10.

An alternative would be to move the benchmark to a separate repo. How important is it that ppxlib continue to support versions of ocaml older than 4.10?

**Comment by pitag-ha** [comment]
> How important is it that ppxlib continue to support versions of ocaml older than 4.10?

Short and concrete answer to this: I'd prefer to avoid dropping compiler support up to 4.10 to add benchmarks.

Longer and more general answer to this: For the same reason @panglesd mentioned above why we never add new hard dependencies to `ppxlib`, we also tend to try to avoid dropping compiler version support for `ppxlib`: a big part of the `opam` ecosystem depends transitively on `ppxlib` and some package maintainers wouldn't be happy if they'd have to drop support for some compilers because of `ppxlib`. That being said, when there's a good reason to drop old compilers, we do so. Usually the hard lower bound is the min of the versions rescript etc use (I think that's currently still lower than 4.10.0).


> This has resulted in ci only being run on versions of the compiler >= 4.10.

As mentioned above, to fix this I think you could have a compiler >=  4.04.1 bound in the `opam` file and desable the builds of the benchmark files in the CI by adding
```
(enabled_if
  (>= %{ocaml_version} "4.10.0"))
```
to their dune files.

**Comment by gridbugs** [comment]
I've updated all the benchmarking dune files so that they only build on ocaml >= 4.10, and now the tests for older version of the compiler are running again.

**Comment by panglesd** [comment]
Thanks! Sorry for leading you into the wrong direction with the package thing.

I'll review the code soon!

File: bench/bench.ml
Comment (panglesd) [inline bench/bench.ml:198]: This code is just to deal with the case where bench is called somewhere else than the root of the project. That's a nice attention to details :+1:

**Review by panglesd — APPROVED** [review]
Looks good to me!

I made a suggestion for a more human-readable output for local benchmarking, otherwise the code, and benchmark organization, looks great! It will be easy to add new benchmarks.

**Comment by gridbugs** [comment]
> Thanks! Sorry for leading you into the wrong direction with the package thing.
> 
> I'll review the code soon!

All good! I still ended up making a ppx-bench package to handle extra dependencies required only for benchmarking.

File: bench/bench.ml
Comment (gridbugs) [reply bench/bench.ml]: Thanks!

File: ppxlib-bench.opam
Comment (panglesd) [inline ppxlib-bench.opam:18]: Could you add this for consistency with the `dune-project` file?

```suggestion
  "base" {>= "v0.15" & < "v0.16"}
```

Comment (gridbugs) [reply ppxlib-bench.opam:18]: I ended up removing the versions from `dune-project` instead because they were causing CI to limit the versions of ocaml used for testing to just those compatible with that range of versions.

**Comment by panglesd** [comment]
Thanks, @gridbugs! Looking forward to integrating this with `current-bench`.

**Comment by panglesd** [comment]
I tried to enrol this repository in ocaml-benchmark, but "This action must be performed by an organization owner".

Also, there are no public people in the `ocaml-ppx` organization, so I don't know who to ping!

**Comment by pitag-ha** [comment]
I already enrolled it on Monday: https://autumn.ocamllabs.io/ocaml-ppx/ppxlib . I think anyone of us could have done it (we should have the same permissions).

There seems to be a bug, though: notice the "No data for selected interval.". We think it's on the `current-bench` side and @ElectreAAS is already kindly looking into it.

**Comment by kit-ty-kate** [comment]
Hi, this change broke all most of the packages relying on mirage. Is there a particular reason why ppx_sexp_conv is vendored? Why not pulling it up as a normal opam dependency of the ppxlib-bench package?

**Comment by panglesd** [comment]
Ouch, sorry for the mess!

I can't remember or find any reason to vendor instead of pulling it as normal dependency. Maybe it was needed before the `ppxlib-bench` package was created? @gridbugs do you remember?

Without reason to do otherwise, I'll open a PR removing the vendored bench dependency, and make a `0.29.1` release quickly.

**Comment by ceastlund** [comment]
Last I spoke to @pitag-ha about it, it was vendored in order to have more control over its versioning. I don't know any deeper than that what the reasoning was, perhaps Sonja remembers.

**Comment by kit-ty-kate** [comment]
I may understand why it was done this way, after trying to patch this. `ppx_sexp_conv` is used as the main user of ppxlib for the benchmark. So you need it to be compiled using ppxlib inside the project. Installing it from opam would use the latest stable instead (and would make the bench target unbuildable)

I'm not sure what's the best way to fix that. So far I'm leaning into reduce the vendor by removing everything that's not used and removing the public libraries, which should fix the issues when vendoring ppxlib itself. I'm having an issue with the extender which i can't get to compile for some reason tho at the moment so i haven't opened a PR yet.

**Comment by pitag-ha** [comment]
Arrg. Thanks for pointing this out, @kit-ty-kate!

We can choose a different PPX for benchmarking and vendor that one (and get rid of `ppx_sexp_conv`). It doesn't need to be `ppx_sexp_conv`. @kit-ty-kate, do you have an idea/suggestion which PPX we could vendor without breaking anyone on `opam`? If not, one of us will have a look through the PPXs to find a good fit. In case there's none (although I'd be surprised), we'll write our custom benchmark PPX.

**Comment by gridbugs** [comment]
If I remember correctly I chose to vendor `ppx_sexp_conv` because I didn't want it to be possible to change the version of `ppx_sexp_conv ` used in the benchmark experiments as this could impact the results of the benchmarks. Apologies for the breakage! Thanks all for sorting it out.

@kit-ty-kate mind sharing why vendoring `ppx_sex_conv` caused such a breakage? Are there potential dangers of vendoring packages?

**Comment by kit-ty-kate** [comment]
@gridbugs i wrote about it in the description of https://github.com/ocaml-ppx/ppxlib/pull/386. If you need more specific information I can add it there

**Comment by gridbugs** [comment]
Excellent, thanks @kit-ty-kate!

**Pull request merged**
